.travis.yml

```diff
@@ -1,227 +1,471 @@
+        - PACKAGING=yes
+        - TESTING=yes
+        - STANDALONE_TESTS_ONLY=no
+      addons:
+        homebrew:
+          packages:
+            - unixodbc
+
+before_install: |-
+  set -Eeo pipefail
+  unalias cd || true
+  unset -f cd || true
+
+  if [[ "$TRAVIS_OS_NAME" == "linux" ]]; then
+    sudo apt-get update -q
+  elif [[ "$TRAVIS_OS_NAME" == "osx" ]]; then
+    cd "$(brew --repo)" && git fetch && git reset --hard origin/master && cd -
+    brew update
+  elif [[ "$TRAVIS_OS_NAME" == "windows" ]]; then
+    : # choco upgrade
+  fi
+
+install: |-
+  if [[ "$TRAVIS_OS_NAME" == "linux" ]]; then
+    sudo apt-get install -y build-essential cmake perl libdbi-perl libdbd-odbc-perl python python-pyodbc
+  elif [[ "$TRAVIS_OS_NAME" == "osx" ]]; then
+    brew uninstall --force --ignore-dependencies cmake perl python@2
+    brew install cmake perl python@2
+    PERL_MM_USE_DEFAULT=1 cpan -T DBD::ODBC
+    if [[ "$ODBC_PROVIDER" == "UnixODBC" ]]; then
+      pip install --user pyodbc
     fi
-  - cmake -G Ninja -DCMAKE_CXX_COMPILER=`which $CXX` -DCMAKE_C_COMPILER=`which $CC` -DTEST_DSN=clickhouse_localhost -DTEST_DSN_W=clickhouse_localhost_w $CMAKE_FLAGS ..
-  - cmake --build . --target all
-  # unit-tests
-  - ctest -R '.*-ut.*'
-  # more complicated tests
-  - |
-    if [ "$TRAVIS_OS_NAME" == "osx" ]; then
-      # install more deps for python & perl to run more tests    
-      sudo PERL_MM_USE_DEFAULT=1 cpan -T DBD::ODBC
-      curl https://bootstrap.pypa.io/get-pip.py | sudo python
-      # it looks like that older version works with iodbc out of the box on Mac
-      # https://github.com/mkleehammer/pyodbc/commit/aa178bfd25afd6fb90c77d054c1bcf0a6aa54411
-      sudo pip install pyodbc==3.0.7
-      # In macOS, ClickHouse is installed and running as a separate process on this same system.
-      CLICKHOUSE_SERVER_IP='localhost'
-      ODBC_DRIVER_PATH=$(realpath driver/libclickhouseodbc.dylib)
-      ODBC_DRIVERW_PATH=$(realpath driver/libclickhouseodbcw.dylib)
-    else 
-      # Using absolute path to compiler to fix setup.py issue of pyodbc: unable to execute 'clang': No such file or directory
-      #- sudo CC=`which $CC` CXX=`which $CXX` pip install pyodbc -V --disable-pip-version-check
-      # To simplify things a bit, start ClickHouse server in docker
-      echo "Clickhouse image ${CLICKHOUSE_SERVER_IMAGE}";
+  elif [[ "$TRAVIS_OS_NAME" == "windows" ]]; then
+    export OLD_VSWHERE=yes
+
+    # TODO: use choco instead of manual download/unpack/env setting, once 3.14+ version
+    # is available, since starting from 3.14 WiX Toolset doesn't depend on .NET 3.5,
+    # which has problems installing via choco.
+    # choco install wixtoolset --yes
+    wget https://wixtoolset.org/downloads/v3.14.0.712/wix314-binaries.zip
+    powershell -Command Expand-Archive wix314-binaries.zip -DestinationPath wix314-binaries
+    export WIX="$(cd wix314-binaries && pwd -W)"
+  fi
+
+.clickhouse_start: &clickhouse_start |-
```

Comment:
I suggest moving refactoring entire `script` section (and its dependencies) by moving it to a separate file: `scripts/build.sh` (or any other name).

That would greatly simplify `.travis.yml`, somewhat simplify the scripts (as it will be pure bash, instead of bash-in-yaml, which has some limitations), isolate environment modifications and so on.

Reply:
Initially, I agreed to your suggestion, and was going to move the scripts, but now, once I started reviewing the script, I realize, that they are pretty separated, and the only one section that is relatively lengthy, is lengthy because of "heredoc" parts of config, the actual bash code is 10-20 lines in each of those 9 sections. Now I have a strong feeling that moving all these steps into separate bash scripts will make it harder to maintain them, as it won't be clear where all those vars come from, for example, and patterns of `if`'s won't be so obvious, and so on.

Reply:
Ok, let's leave as it is and see whether it is practical or not.